File: examples/01-foundations/go/deterministic_replay_bad.go
package foundations

import (
	"math/rand"
	"os"
	"strconv"
	"time"

	"go.temporal.io/sdk/workflow"
)

func BadReplayWorkflow(ctx workflow.Context, batchDate string) error {
	// Bad: replay re-runs Workflow code. This value changes on replay.
	now := time.Now()

	// Bad: this random value is not recorded in Workflow history.
	shard := rand.Intn(10)

	// Bad: direct I/O in Workflow code can run again during replay.
	return os.WriteFile("/tmp/workflow.log",
		[]byte(strconv.FormatInt(now.UnixMilli(), 10)+":"+strconv.Itoa(shard)), 0o644)
}

File: examples/01-foundations/python/deterministic_replay_bad.py
import random
import time
from datetime import datetime

from temporalio import workflow


@workflow.defn
class BadReplayWorkflow:
    @workflow.run
    async def run(self, batch_date: str) -> None:
        # Bad: replay re-runs Workflow code. This value changes on replay.
        now = time.time()

        # Bad: this random value is not recorded in Workflow history.
        shard = random.randint(0, 9)

        # Bad: direct I/O in Workflow code can run again during replay.
        with open("/tmp/workflow.log", "w") as f:
            f.write(f"{now}:{shard}")
        # (datetime.now() is just as unsafe — never read wall-clock here.)
        _ = datetime.now()
